text/0000-repr-type-aliases.md

```diff
@@ -0,0 +1,111 @@
+
+To aid macro authors, an allow-by-default lint should be added that warns a user if they use `discriminant = ...` with a relative path instead of an absolute one, since the resolved type depends on the existing scope. For example, `discriminant = u32` or `discriminant = my_type` could potentially be shadowed by another type in the scope, but `discriminant = ::std::u32` and `discriminant = $crate::my_type` will always work as expected. This could potentially be relegated to a `clippy::pedantic` lint instead of a `rustc` lint.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The requirement for `discriminant =` is unfortunate, but it feels like the best way to ensure that adding new representations isn't a breaking change going forward. Even if we were to decide it weren't a "breaking change," it would still break things anyway, being de-facto breaking. Some people have expressed desire for a dedicated syntax instead, which would fix this problem, but this can still be added after the fact.
+
+Some proc macros like [`zerocopy::FromBytes`] may also have to complicate their logic if they depend on checking the existing `#[repr]` attribute for validity. In particular, since they can no longer exactly know which primitive type is being used for the representation, they would have to instead depend on associated constants on the types like `BITS` to verify which type is being used. Instead of being able to emit an error at macro expansion time, this error will have to be triggered at constant evaluation time or runtime instead, which is unfortunate.
+
+Complicating proc macros in this way is considered acceptable for the following reasons:
+
+* Since errors can still be included in constants now that `const` panics are stable, errors can still be caught at compile time.
+* Users of special `repr` attributes are more likely to have more knowledge about specifics of how these proc macros work, and are thus more equipped to deal with weird errors like these.
+* Ultimately, the benefits of adding the feature outweigh this negative.
+
+[`zerocopy::FromBytes`]: https://docs.rs/zerocopy/latest/zerocopy/derive.FromBytes.html
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could always not do this.
+
+But more realistically, here are some alternative designs that were rejected.
+
+## `type`
+
+The second proposal for this RFC used `type = ...` instead of `discriminant = ...`. Initially, this decision was chosen because `discriminant` was long to type, but it was ultimately decided that `discriminant` is more clear and that the extra typing is worth it. Additionally, RFC [#3607] proposes using `discriminant` in its proposed syntax, so, this would be further in line with that as well.
```

Comment:
Since we intend for `repr(discriminant = u8)` to be equivalent to `repr(u8)`, then I'd like to propose that we perhaps *shouldn't* use `repr(discriminant = ...)` since it might complicate a future possibility for enums.

That future possibility — alluded to in this RFC — is that we someday might want to expand the accepted set of discriminant types of enums (e.g., to include things implementing `StructuralEq`). And, if we had such functionality, it might also be nice to declare up front what the discriminant type is. A natural syntax for this could be `#[repr(discriminant = ...)]`.

*However*, `repr(discriminant = ...)`, as proposed by this RFC, does not only set the discriminant of an enum, but *also* affects its memory layout. This is not desirable for all enums. If this RFC is adopted as-is, we'd need to find an alternative way to say "the discriminant type of this enum is X, but I don't care what the enum's in-memory representation is".

Given this, I'd like to suggest that this RFC consider `repr(tag = ...)`. Doing so has three advantages:
1. it makes it clearer that the annotation has an affect on the enum's in-memory representation, not just its logical representation
2. it keeps `repr(discriminant = ...)` syntactically reserved for the aforementioned future possibility in which the user explicitly provides the type of the discriminant
3. it's significantly shorter than "discriminant"

If not `repr(tag = ...)` than perhaps something else.

Reply:
I'm confused how labeling this a tag makes it less confusing. It's a shorter word, but that's about it.

We've already established that we're changing the memory layout by using `repr`.

Reply:
Considering that these are often called "tagged unions" in contexts where memory layout matters, I do think the name is clearer. It's also going to be consistent with other in-flight changes to Rust's documentation: https://github.com/rust-lang/reference/pull/1454#discussion_r1671867991

What are the advantages of `discriminant`?

Reply:
They are often called this, but we also explicitly use the term discriminant in places where it has an API meaning. See: https://doc.rust-lang.org/nightly/std/mem/fn.discriminant.html

Reply:
In the lang meeting today, we decided on `tag`.  We felt that the existing `mem::discriminant` API was unfortunate, and unloved in enough other ways, that we didn't need to weigh that heavily as precedent.  See also [here](https://github.com/rust-lang/rfcs/pull/3659#issuecomment-2221229758).

Reply:
Very fair. I think that it would be worth proposing to deprecate & rename that API if this is the route we're going, to ensure uniformity.

Reply:
(Also, while the decision on the rename is resolved, I'm going to unresolve this until I update the RFC to rename the attribute, just so it's easier to track.)